Add an introduction to the generative tests (#1159)
opensafely-core/databuilder: tests/generative/README.md

    bool: st.booleans(),
    datetime.date: st.dates(
        min_value=datetime.date(2010, 1, 1), max_value=datetime.date(2020, 12, 31)
    ),
    float: st.floats(min_value=0.0, max_value=11.0, width=16, allow_infinity=False),
    str: st.text(alphabet=["a", "b", "c"], min_size=0, max_size=3),
}
```

The same value strategies are used for defining both data strategies and variable (query model)
strategies. Generally, we try to define these values with narrow enough ranges that they will
sometimes overlap, and we can test equality. e.g. if we are testing an addition operation that
takes 2 ints, we can be reasonably sure that `st.integers(min_value=0, max_value=10)` will
test addition of two ints that are the same at some point.

## variable strategies

The variable strategies are the most complex part of databuilder's generative test strategies.

A variable is defined by calling `variable()` in [`variable_strategies.py`](variable_strategies.py),
with the tables, schema and value strategies as described above.

`variable` defines lots of inner functions, each of which returns a strategy for creating the
thing it is named for, not the thing itself.

> 

Reviewer: An example would be useful.

```suggestion
For example, the `value` inner function returns a strategy for creating `Value` objects, and not a `Value` object itself.

```

Within `series()`, we define `series_constraints`; this is a dict of all possible strategies for
operations that produce a series. The keys are strategy callables. The values are 2-tuples representing the possible return types of the series that will be generated by that strategy,
and the possible domains of the frame it can be drawn from.

Frame domains indicate whether a particular `Series` node needs to be drawn from a patient table (i.e.one-row-per-patient), a non-patient table, or either.

The `series()` strategy chooses from the possible operation strategies that meet the constraints of
the `type_` and `frame` passed into it.

### A simple example: count

For example, if we only had strategies for `exists` and `count`, and we call `series(int, patient_table)`,
where `patient_table` has been chosen via `one_row_per_patient_frame()`:
```
    def series(type_, frame):
        ...
        # define contraints for possible series strategies
        series_constraints = {
            exists: ({bool}, DomainConstraint.PATIENT),
            count: ({int}, DomainConstraint.PATIENT),
        }
        ...
        def constraints_match(s):
            ...

>         # find possible series stratgeies that match contraints, and choose one

Reviewer 2: ```suggestion
        # find possible series strategies that match constraints, and choose one
```

# Generative Tests Overview

The generative (property-based) tests use Hypothesis to generate query model variable definitions.

There are detailed docstrings in the code, and a description of how to run the tests in the
[developer docs](../../DEVELOPERS.md#generative-tests). This document is an attempt to
provide an overview and introduction to what the tests are doing.

There is one main test, `test_query_model` in [test_query_model.py](test_query_model.py). This
generates a variable definition, executes it using the MSSQL, SQLite and in-memory query engines,
and checks that the results are the same.

Note that because the inputs are automatically generated, we can't test that the results are
*correct*, but we rely on other (unit, integration, acceptance) tests to check the exact outputs
produced by the query engines, and we assume that the engines are sufficiently different in
implementation that it is a sufficient test to check that they always produce the same results
from the same inputs.

## Terminology

We are used to thinking about query model things such as `Series` as representing a concrete column

> in a table. The generative tests require a mental shift from thinking in terms of `strategies` - i.e.

Author: I've reworded this.  I only wrote half the sentence I'd intended to write :)

A variable is defined by calling `variable()` in [`variable_strategies.py`](variable_strategies.py),
with the tables, schema and value strategies as described above.

`variable` defines lots of inner functions, each of which returns a strategy for creating the
thing it is named for, not the thing itself.
For example, the `value` inner function returns a strategy for creating `Value` objects, and not a `Value` object itself.

A valid variable is a `Series`, chosen by selecting a type (one of the types in `value_strategies`) and a frame
that the variable must be consistent with (with one row per patient, because we require that any
variable on a dataset represents one row per patient). The variable returned will be a
series of the chosen type, consistent with the chosen frame.

```
def variable(patient_tables, event_tables, schema, value_strategies):
   ...

    @st.composite
    def valid_variable(draw):
        type_ = draw(any_type())
        frame = draw(one_row_per_patient_frame())
        return draw(series(type_, frame))

    return valid_variable()
```

Let's take a closer look at the elements of the `valid_variable`.

> `type_` is a strategy for choosing a type, sampled from the types used as the keys for

Reviewer 3: This isn't quite right. `any_type()` returns the strategy you describe. `type_` is an actual type, drawn from the strategy. (The same applies to `frame` below.)

```
    def series(type_, frame):
        ...
        # define contraints for possible series strategies
        series_constraints = {
            exists: ({bool}, DomainConstraint.PATIENT),
            count: ({int}, DomainConstraint.PATIENT),
        }
        ...
        def constraints_match(s):
            ...

        # find possible series strategies that match constraints, and choose one
        possible_series = [s for s in series_types if constraints_match(s)]
        series_strategy = draw(st.sampled_from(possible_series))

        # draw a series from the chosen strategy
        return draw(series_strategy(type_, frame))

```
`exists` returns a bool series with one row per patient.

`count` returns an int series with one row per patient.

The `frame` we've passed in matches the domain constriants for both strategies (it's a
one-row-per-patient frame). However, `exists` produces a bool series, and we need a strategy
that produces an int series, so `count` is the only possible strategy that matches. `series()`
will draw from the `count` strategy.

Note that the `type_` and `frame` are always passed on as arguments to the selected series

> strategy, and each individual series operation strategy has the same function signature.

Reviewer 3: ```suggestion
strategy, so that all series operation strategies have the same function signature.
```

Maybe slightly clearer?
